smart-contracts/contracts/Unlock.sol

```diff
@@ -509,6 +528,112 @@ contract Unlock is UnlockInitializable, UnlockOwnable {
     }
   }
 
+  function getBalance(address token) internal view returns (uint) {
+    return token == address(0) ?
+      address(this).balance 
+      :
+      IERC20(token).balanceOf(address(this));
+  }
+
+  /**
+   * Swap tokens and call a function in a lock contract
+   * @notice If the actual amount spent is less than the specified maximum amount, the remaining tokens wll 
+   * be held by the Unlock contract
+   */
+  function swapAndCall(
+    address lock,
+    address srcToken,
+    uint amountInMax,
+    address swapRouter,
+    bytes memory swapCalldata,
+    bytes memory callData
+  ) public payable returns(bytes memory) {
+    // check if lock exists
+    if(!locks[lock].deployed) {
+      revert('l');
+    }
+
+    // get lock pricing 
+    address destToken = IPublicLock(lock).tokenAddress();
+
+    // get price in destToken from lock
+    uint keyPrice = IPublicLock(lock).keyPrice();
+
+    // get balances of Unlock before
+    // if payments in ETH, substract the value sent by user to get actual balance
+    uint balanceTokenDestBefore = destToken == address(0) ? 
+      getBalance(destToken) - msg.value 
+            :
+      getBalance(destToken);
+
+    uint balanceTokenSrcBefore = 
+        srcToken == address(0) ? 
+          getBalance(srcToken) - msg.value 
+          :
+          getBalance(srcToken);
+
+    if(srcToken != address(0)) {
+      // Transfer the specified amount of ERC20 to this contract
+      TransferHelper.safeTransferFrom(srcToken, msg.sender, address(this), amountInMax);
+
+      // Approve the router to spend ERC20.
+      TransferHelper.safeApprove(srcToken, swapRouter, amountInMax);
+    }
+
+    // calculate value to send to Uniswap
+    uint swapValue = srcToken == address(0) ? msg.value : 0;
+
+    // executes the swap
+    (bool success,) = swapRouter.call{ value: swapValue }(swapCalldata);
+    // make sure Uniswap revert
+    if(success == false) {
```

Comment:
Not sure we need this if have the balance check from below?

Reply:
yes, we need to propagate possible errors from uniswap